pkg/instancetype/controller/vm/controller_test.go

```diff
@@ -0,0 +1,1052 @@
+				Name: instancetypeObj.Name,
+				Kind: instancetypeapi.SingularResourceName,
+			}
+
+			vm.Spec.Template.Spec.Domain.CPU = &virtv1.CPU{
+				Sockets: uint32(1),
+				Cores:   uint32(4),
+				Threads: uint32(1),
+			}
+
+			_, err := instancetypeController.Sync(vm, vmi)
+			Expect(err).To(MatchError(ContainSubstring("conflicts with selected instance type")))
+			testutils.ExpectEvents(recorder, common.FailedCreateVirtualMachineReason)
+		})
+
+		It("should fail to sync if an existing ControllerRevision is found with unexpected VirtualMachineInstancetypeSpec data", func() {
+			unexpectedInstancetype := instancetypeObj.DeepCopy()
+			unexpectedInstancetype.Spec.CPU.Guest = 15
+
+			instancetypeRevision, err := instancetype.CreateControllerRevision(vm, unexpectedInstancetype)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(
+				context.Background(), instancetypeRevision, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+				Name: instancetypeObj.Name,
+				Kind: instancetypeapi.SingularResourceName,
+			}
+
+			_, err = instancetypeController.Sync(vm, vmi)
+			Expect(err).To(MatchError(ContainSubstring("found existing ControllerRevision with unexpected data")))
+			testutils.ExpectEvents(recorder, common.FailedCreateVirtualMachineReason)
+		})
+	})
+
+	Context("preference", func() {
+		var clusterPreference *v1beta1.VirtualMachineClusterPreference
+
+		BeforeEach(func() {
+			clusterPreference = &v1beta1.VirtualMachineClusterPreference{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "clusterPreference",
+					UID:        resourceUID,
+					Generation: resourceGeneration,
+				},
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: v1beta1.SchemeGroupVersion.String(),
+					Kind:       "VirtualMachineClusterPreference",
+				},
+				Spec: v1beta1.VirtualMachinePreferenceSpec{
+					Firmware: &v1beta1.FirmwarePreferences{
+						DeprecatedPreferredUseEfi: pointer.P(true),
+					},
+					Devices: &v1beta1.DevicePreferences{
+						PreferredDiskBus:        virtv1.DiskBusVirtio,
+						PreferredInterfaceModel: "virtio",
+						PreferredInputBus:       virtv1.InputBusUSB,
+						PreferredInputType:      virtv1.InputTypeTablet,
+					},
+				},
+			}
+			_, err := virtClient.VirtualMachineClusterPreference().Create(context.Background(), clusterPreference, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			err = clusterPreferenceInformerStore.Add(clusterPreference)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should store VirtualMachinePreference as ControllerRevision on Sync", func() {
+			vm.Spec.Preference = &virtv1.PreferenceMatcher{
+				Name: preference.Name,
+				Kind: instancetypeapi.SingularPreferenceResourceName,
+			}
+
+			var err error
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Create(context.TODO(), vm, metav1.CreateOptions{})
+			Expect(err).To(Succeed())
```

Comment:
Would it be possible to unify the error checking?
`Expect(err).To(Succeed())` vs `Expect(err).ToNot(HaveOccurred())`